srml/support/src/storage/hashed/generator.rs

```diff
@@ -0,0 +1,244 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Abstract storage to use on Twox128Storage trait
+
+use crate::codec;
+use crate::rstd::prelude::{Vec, Box};
+#[cfg(feature = "std")]
+use crate::storage::unhashed::generator::UnhashedStorage;
+#[cfg(feature = "std")]
+use runtime_io::{twox_128, blake2_128};
+
+/// Abstraction around storage.
+pub trait Twox128Storage {
+	/// true if the key exists in storage.
+	fn exists(&self, key: &[u8]) -> bool;
+
+	/// Load the bytes of a key from storage. Can panic if the type is incorrect.
+	fn get<T: codec::Decode>(&self, key: &[u8]) -> Option<T>;
+
+	/// Load the bytes of a key from storage. Can panic if the type is incorrect. Will panic if
+	/// it's not there.
+	fn require<T: codec::Decode>(&self, key: &[u8]) -> T { self.get(key).expect("Required values must be in storage") }
+
+	/// Load the bytes of a key from storage. Can panic if the type is incorrect. The type's
+	/// default is returned if it's not there.
+	fn get_or_default<T: codec::Decode + Default>(&self, key: &[u8]) -> T { self.get(key).unwrap_or_default() }
+
+	/// Put a value in under a key.
+	fn put<T: codec::Encode>(&self, key: &[u8], val: &T);
+
+	/// Remove the bytes of a key from storage.
+	fn kill(&self, key: &[u8]);
+
+	/// Take a value from storage, deleting it after reading.
+	fn take<T: codec::Decode>(&self, key: &[u8]) -> Option<T> {
+		let value = self.get(key);
+		self.kill(key);
+		value
+	}
+
+	/// Take a value from storage, deleting it after reading.
+	fn take_or_panic<T: codec::Decode>(&self, key: &[u8]) -> T { self.take(key).expect("Required values must be in storage") }
+
+	/// Take a value from storage, deleting it after reading.
+	fn take_or_default<T: codec::Decode + Default>(&self, key: &[u8]) -> T { self.take(key).unwrap_or_default() }
+}
+
+// We use a construct like this during when genesis storage is being built.
+#[cfg(feature = "std")]
+impl Twox128Storage for crate::rstd::cell::RefCell<&mut sr_primitives::StorageOverlay> {
+	fn exists(&self, key: &[u8]) -> bool {
+		UnhashedStorage::exists(self, &twox_128(key))
+	}
+
+	fn get<T: codec::Decode>(&self, key: &[u8]) -> Option<T> {
+		UnhashedStorage::get(self, &twox_128(key))
+	}
+
+	fn put<T: codec::Encode>(&self, key: &[u8], val: &T) {
+		UnhashedStorage::put(self, &twox_128(key), val)
+	}
+
+	fn kill(&self, key: &[u8]) {
+		UnhashedStorage::kill(self, &twox_128(key))
+	}
+}
+
+/// Abstraction around storage.
+pub trait Blake2_128Storage {
```

Comment:
Do we really need to copy this?

Can we not create the following:

```
pub trait Storage<Hash> {
    //blabla
}

impl Storage<Blake2> for RuntimeStorage {}

impl Storage<Twox> for RuntimeStorage {}
```

And if you still want aliases for the storage traits, you could probably do the following:
```
pub trait Blake2_128Storage: Storage<Blake2> {}
```

Reply:
yes, I can introduce local structures Blake2 and Twox or I can make use of Hasher in primitives.
currently only Blake2Hasher is there.

> also in generator we could refactor a bit by introducing HashedStorage<H: Hasher> and create Hasher structs for Twox128 and Blake2_128 in core/primitives (because only Blake2Hasher (blake2_256) exists for now).

I think we can go with local structures

Reply:
addressed.

also I implement StorageValue generator for Twox only at the moment and StorageMap for Blake2 only.

Do we want to allow generator such as StorageMap using twox ? this can be done easily now.

Also make decl_storage configurable could be done. But maybe in another PR ?